One-shot tasks

furquan-rivos opened the pull request:
Enable support for tasks to move to stopped state. In the future, we will need changes to enable some task (supervisor/controller) to decide that we are done with all work and ask kernel to stop by doing a wfi or jumping to some other entity. This is the first step towards that change.

furquan-rivos committed:
abi: Add Kipcnum for ExitCurrentTask

This change adds a new kernel ipc message for exiting current task. This can be
used by any task to move itself into stopped state. This is required for
introducing one-shot type of tasks in hubris.

furquan-rivos committed:
kernel: Handle newly introduced kipc for exiting current task

This change adds support in kernel to handle the kipc for moving current task to
stopped state. This path is treated similar to the task fault path to notify the
supervisor task about the change in task state. This allows supervisor to take
any appropriate action as required.

furquan-rivos committed:
userlib: Add a helper function for stopping current task

This change adds a helpr function `exit_current_task` that can be used by
one-shot tasks to exit once it is done with its work.

furquan-rivos committed:
task: Add a sample one-shot task

This task doesn't really do anything much interesting apart from printing out a
hello message and exiting by calling `kipc::exit_current_task`.

furquan-rivos committed:
demo-hifive-inventor: Add one-shot task

This change adds the one-shot task to demo-hifive-inventor to demonstrate the
one-shot behavior.

TEST=Verified using humility that the state reported for one_shot task
is as expected:

humility: attached via QEMU GDB server
system time = 36
ID TASK                 GEN PRI STATE
 0 jefe                   0   0 recv, notif: bit0 bit1(T+65)
 1 ext_int_ctrl           0   1 recv, notif: bit0(irq11)
 2 rtc                    0   2 RUNNING
 3 pong                   0   3 ready
 4 ping                   0   4 ready
 5 one_shot               0   1 Stopped
 6 idle                   0   5 ready

furquan-rivos committed:
Regenerate Cargo.lock

This change regenerates Cargo.lock after adding one-shot task to
demo-hifive-inventor app.

furquan-rivos committed:
Run cargo fmt

furquan-rivos committed:
kernel: Rename HUBRIS_FAULT_NOTIFICATION to HUBRIS_TASK_STATE_CHANGE_NOTIFICATION

This change renames HUBRIS_FAULT_NOTIFICATION to
HUBRIS_TASK_STATE_CHANGE_NOTIFICATION since this notification is sent to
supervisor task not only in case of faults but also when task moves to stopped
state.

furquan-rivos committed:
task/jefe: Rename FAULT_MASK to TASK_STATE_CHANGE_MASK

Similar to kernel change, rename FAULT_MASK to TASK_STATE_CAHNGE_MASK to better
capture the notification type.

furquan-rivos committed:
kernel: Move HUBRIS_TASK_STATE_CHANGE_NOTIFICATION to task.rs

This is accessed only in task.rs and so there is no need to make it public and
define in startup.rs. Hence, this change moves the definition of
HUBRIS_TASK_STATE_CHANGE_NOTIFICATION to task.rs.

furquan-rivos commented:
[This humility change](https://github.com/rivosinc/humility/pull/13) goes along with this hubris change.

timw-rivos commented on task/one-shot/README.mkdn:
```diff
@@ -0,0 +1,3 @@
+# Task One shot
+
```
Long line, wrap at 96 chars?

timw-rivos commented on task/one-shot/src/main.rs:
```diff
@@ -0,0 +1,18 @@
+    kipc::exit_current_task();
+
+    // Adding this just to make rustc happy
+    loop {}
```
is there an equivalent of `unreachable` in Rust?

furquan-rivos replied:
Thanks! That is what I needed. Updated in latest version.

furquan-rivos committed:
task/one-shot: Replace loop{} with unreachable!()

This change uses unreachable! to indicate that the call to exit_current_task is
not expected to return.

furquan-rivos committed:
task/one-shot: Wrap README.mkdn at 96 characters

adurbin-rivos commented on sys/abi/src/lib.rs:
```diff
@@ -605,6 +605,7 @@ pub enum Kipcnum {
     FaultTask = 3,
     ReadImageId = 4,
     Reset = 5,
+    ExitCurrentTask = 6,
```
Exit is not really an exit, though? It's a suspended/stopped task? Does the task start executing again where it was suspended/stopped? Or is it a clean restart?

furquan-rivos replied:
> Exit is not really an exit, though?

Exit is not an exit of the kernel, it is basically moving the task to stopped state. I can rename it to StopCurrentTask if that seems more appropriate.

> Does the task start executing again where it was suspended/stopped? Or is it a clean restart?

It is a clean start at the entrypoint if the supervisor task calls a restart task for the stopped task.

adurbin-rivos replied:
I was trying to understand the lifecycle of the task based on the ipcs. I know those are likely not 1:1. When a task is in the stopped state can it only be restarted?

furquan-rivos replied:
Correct. It cannot receive interrupts or notifications or be brought back to where it decided to exit. Instead the only way to bring the task back is the restart system call made by another task.

furquan-rivos replied:
I left the exit name as is and added a new exited state to indicate that the task ran and exited.

adurbin-rivos commented on sys/kern/src/task.rs:
```diff
@@ -846,13 +856,24 @@ pub fn force_fault(
+    tasks: &mut[Task],
+    index: usize,
+) -> NextTask {
+    tasks[index].set_healthy_state(SchedState::Stopped);
```
Forgive my uninformed questions, but from what I can gather this function is called from user context. So the user context has direct access to tasks' state?

furquan-rivos replied:
User context makes a library call [`exit_current_task`](https://github.com/rivosinc/hubris/blob/d8780f05af6f4c7b93f890153f3f5cfa67d7c7e2/sys/userlib/src/kipc.rs#L27) which ends up making a syscall [`sys_send`](https://github.com/rivosinc/hubris/blob/d8780f05af6f4c7b93f890153f3f5cfa67d7c7e2/sys/userlib/src/kipc.rs#L28) to the kernel. Within the kernel, it identifies that it is a message directed towards itself and not a different task and so the control goes to [`handle_kernel_message`](https://github.com/rivosinc/hubris/blob/d8780f05af6f4c7b93f890153f3f5cfa67d7c7e2/sys/kern/src/kipc.rs#L16) which ends up calling this `exit_task` function. So, user context does not have direct access to task state. It instead makes a syscall to request kernel to update the state.

adurbin-rivos replied:
Got it. I think I was confused which exit_current_task() was being invoked from the user code.

beezow commented on sys/kern/src/task.rs:
```diff
@@ -809,6 +810,16 @@ pub fn priority_scan(
 
+fn notify_supervisor(tasks: &mut [Task]) -> NextTask {
+    let supervisor_awoken =
+        tasks[0].post(NotificationSet(HUBRIS_TASK_STATE_CHANGE_NOTIFICATION));
```
Does this assume the supervisor is task 0?  Is this guaranteed?

furquan-rivos replied:
That is already the assumption in hubris. I just moved the code that was part of the `force_fault` function into its own function so that it can be used in `exit_task` as well.

beezow replied:
Yea, I saw that it was just moved, so I don't think we need to change it.  Was more of a curiosity.

furquan-rivos replied:
[Hubris reference documentation](https://hubris.oxide.computer/reference/#_what_is_the_supervisor) says that "The kernel can spot the supervisor because the supervisor always has task index 0, and is listed first in the app.toml"

beezow reviewed (dismissed):
The changes look good, but I am not sure if I understand why this is needed.  What is the benefit of this over doing a `sys_recv_closed(&mut [], 0, TaskId::KERNEL)`?  It also ends the task as it is impossible to exit from, and displays appropriately in humility.

furquan-rivos commented:
> The changes look good, but I am not sure if I understand why this is needed. What is the benefit of this over doing a `sys_recv_closed(&mut [], 0, TaskId::KERNEL)`? It also ends the task as it is impossible to exit from, and displays appropriately in humility.

With `sys_recv_closed` the task basically gets into a point where it is waiting for nothing `0` from TaskId::KERNEL. Logically, it achieves similar result, but the state of the task is `waiting in recv` and not stopped. I introduced this `exit` as a cleaner way to enter stopped state. It can be utilized by supervisor/controller task to determine what all work is completed by other tasks.

furquan-rivos commented:
> ... displays appropriately in humility.

Curious what you are thinking about humility?

beezow commented:
> > ... displays appropriately in humility.
> 
> Curious what you are thinking about humility?

In humility it displays as `DEAD`.

furquan-rivos commented:
> > > ... displays appropriately in humility.
> > 
> > 
> > Curious what you are thinking about humility?
> 
> In humility it displays as `DEAD`.

humility also has a comment that considers this state as a bug. I am not sure if we really want to overload the state that is considered bug to also represent a state where task decides to cleanly exit.

FawazTirmizi reviewed (dismissed):
> > The changes look good, but I am not sure if I understand why this is needed. What is the benefit of this over doing a `sys_recv_closed(&mut [], 0, TaskId::KERNEL)`? It also ends the task as it is impossible to exit from, and displays appropriately in humility.
> 
> With `sys_recv_closed` the task basically gets into a point where it is waiting for nothing `0` from TaskId::KERNEL. Logically, it achieves similar result, but the state of the task is `waiting in recv` and not stopped. I introduced this `exit` as a cleaner way to enter stopped state. It can be utilized by supervisor/controller task to determine what all work is completed by other tasks.

In that case, maybe this PR should also have the `sys_recv_closed(&mut [], 0, TaskId::KERNEL)` put the task in the same state? Otherwise, this looks fine to me.

beezow commented:
> > The changes look good, but I am not sure if I understand why this is needed. What is the benefit of this over doing a `sys_recv_closed(&mut [], 0, TaskId::KERNEL)`? It also ends the task as it is impossible to exit from, and displays appropriately in humility.
> 
> With `sys_recv_closed` the task basically gets into a point where it is waiting for nothing `0` from TaskId::KERNEL. Logically, it achieves similar result, but the state of the task is `waiting in recv` and not stopped. I introduced this `exit` as a cleaner way to enter stopped state. It can be utilized by supervisor/controller task to determine what all work is completed by other tasks.

Okay, So this makes it provides a way for the task to notify the supervisor when it stops, without having to explicitly `sys_send` to the supervisor task, right?

beezow commented:
Should we use a separate state for stopped vs not started?  With this change we can no longer determine if a task ran to completion, or if it never started at all.

furquan-rivos commented:
> > > The changes look good, but I am not sure if I understand why this is needed. What is the benefit of this over doing a `sys_recv_closed(&mut [], 0, TaskId::KERNEL)`? It also ends the task as it is impossible to exit from, and displays appropriately in humility.
> > 
> > 
> > With `sys_recv_closed` the task basically gets into a point where it is waiting for nothing `0` from TaskId::KERNEL. Logically, it achieves similar result, but the state of the task is `waiting in recv` and not stopped. I introduced this `exit` as a cleaner way to enter stopped state. It can be utilized by supervisor/controller task to determine what all work is completed by other tasks.
> 
> In that case, maybe this PR should also have the `sys_recv_closed(&mut [], 0, TaskId::KERNEL)` put the task in the same state? Otherwise, this looks fine to me.

The `sys_recv_closed` looks like an unexpected call to kernel if notification bit is set to 0. See the comment in humility where it considers this as a bug. I don't think we should move the tasks to the same state in that case.

furquan-rivos commented:
> > > The changes look good, but I am not sure if I understand why this is needed. What is the benefit of this over doing a `sys_recv_closed(&mut [], 0, TaskId::KERNEL)`? It also ends the task as it is impossible to exit from, and displays appropriately in humility.
> > 
> > 
> > With `sys_recv_closed` the task basically gets into a point where it is waiting for nothing `0` from TaskId::KERNEL. Logically, it achieves similar result, but the state of the task is `waiting in recv` and not stopped. I introduced this `exit` as a cleaner way to enter stopped state. It can be utilized by supervisor/controller task to determine what all work is completed by other tasks.
> 
> Okay, So this makes it provides a way for the task to notify the supervisor when it stops, without having to explicitly `sys_send` to the supervisor task, right?

This `exit` call informs that kernel that the task is done with its work and the kernel then informs the supervisor task. So, it ensures that the task state managed by kernel is also updated. So, yes this wouldn't require any explicit calls from task to supervisor as kernel informs the supervisor of task state change.

furquan-rivos commented:
> Should we use a separate state for stopped vs not started? With this change we can no longer determine if a task ran to completion, or if it never started at all.

I chose `SchedState::Stopped` because it allows us to reuse the same support that already exists to perform restart of tasks [via humility](https://github.com/rivosinc/hubris/blob/9eb39ec7e4e1759a0c2a047173816c74778f2491/task/jefe/src/main.rs#L240) too.

FawazTirmizi commented:
> > Should we use a separate state for stopped vs not started? With this change we can no longer determine if a task ran to completion, or if it never started at all.
> 
> I chose `SchedState::Stopped` because it allows us to reuse the same support that already exists to perform restart of tasks [via humility](https://github.com/rivosinc/hubris/blob/9eb39ec7e4e1759a0c2a047173816c74778f2491/task/jefe/src/main.rs#L240) too.

I don't think this benefit outweighs the risks coming from a lack of distinction between a stopped vs not started state. I'm in favor of adding a new state.

furquan-rivos committed:
Introduce a new state for tasks that run and exit

Based on review feedback, this change introduces a new task state that indicates
that the task ran and exited (SchedState::Exited) so that it is possible to
identify which tasks never ran (and hence in state SchedState::Stopped) v/s
which tasks ran and exited (and hence in state SchedState::Exited).

With this and corresponding humility change, the output reported for tasks is:

humility: attached via QEMU GDB server
system time = 32
ID TASK                 GEN PRI STATE
 0 jefe                   0   0 recv, notif: bit0 bit1(T+68)
 1 ext_int_ctrl           0   1 recv, notif: bit0(irq11)
 2 rtc                    0   2 RUNNING
 3 pong                   0   3 ready
 4 ping                   0   4 ready
 5 one_shot               0   1 EXITED
 6 idle                   0   5 ready

furquan-rivos commented:
> > > Should we use a separate state for stopped vs not started? With this change we can no longer determine if a task ran to completion, or if it never started at all.
> > 
> > 
> > I chose `SchedState::Stopped` because it allows us to reuse the same support that already exists to perform restart of tasks [via humility](https://github.com/rivosinc/hubris/blob/9eb39ec7e4e1759a0c2a047173816c74778f2491/task/jefe/src/main.rs#L240) too.
> 
> I don't think this benefit outweighs the risks coming from a lack of distinction between a stopped vs not started state. I'm in favor of adding a new state.

Ok. Pushed a follow-up change to introduce a separate state.

furquan-rivos committed:
Run cargo fmt

furquan-rivos committed:
Merge branch 'rivos/main' into dev/furquan/one-shot-tasks
